Fixes #46 - Add Flask blueprint handler for GitHub webhooks + helper methods

laghee opened the pull request.

miketaylr commented:
Tests are failing for this (tiny) linting error:

```
./ochazuke/webhooks/helpers.py:201:1: W293 blank line contains whitespace
Exited with code 1
```

miketaylr commented on ochazuke/webhooks/__init__.py:
```diff
@@ -0,0 +1,138 @@
+    By default, we return a 403 HTTP response.
+    """
+    if not is_github_hook(request):
+        return ('Move along, nothing to see here', 401, {'Content-Type':
```
We use `{'Content-Type': 'text/plain'}` quite a bit in this file, perhaps we can make a constant and re-use it? 

`TEXT_PLAIN = {'Content-Type': 'text/plain'}`

miketaylr commented on ochazuke/webhooks/__init__.py:
```diff
@@ -0,0 +1,138 @@
+    if not is_github_hook(request):
+        return ('Move along, nothing to see here', 401, {'Content-Type':
+                                                         'text/plain'})
+    payload = json.loads(request.data)
```
I guess it's probably not very likely, but perhaps we should handle the case where `json.loads()` fails here... like GitHub sent us an empty string or something. Maybe a try/except returning a 403 if it throws?

miketaylr replied:
Or maybe `is_github_hook` could check that the response is something we can parse? Once we get past that, we're good to assume we can operate it.

miketaylr commented on ochazuke/webhooks/__init__.py:
```diff
@@ -0,0 +1,138 @@
+                        {'Content-Type': 'text/plain'})
+        else:
+            # We acknowledge receipt even if we don't process all event types.
+            return ('We\'ll just circular-file that, but thanks!', 202,
```
😄

miketaylr commented on ochazuke/webhooks/helpers.py:
```diff
@@ -0,0 +1,235 @@
+    """Create a hmac comparison.
+
+    Approximates hmac.compare_digest (Py 2.7.7+) until we upgrade.
+    TODO: SEE IF THIS NEEDS TO BE ADAPTED FOR PY3.6
```
Maybe we can file a bug for this? Unless it's a leftover comment?

laghee replied:
@miketaylr This is weird. At first I thought it was a rebase snafu, but I opened the branch locally and it's not there. So I think it's just an earlier commit (there are 5 in here) of the same file? (Same for the secret key stuff below.)

miketaylr commented on ochazuke/webhooks/helpers.py:
```diff
@@ -0,0 +1,235 @@
+        return False
+    post_signature = request.headers.get('X-Hub-Signature')
+    if post_signature:
+        # TODO: DETERMINE WHERE TO HIDE SECRET KEY
```
We should be reading the secret key out of a (Heroku) envvar, I think. A comment here saying as much would be helpful!

laghee replied:
^^

miketaylr commented on ochazuke/webhooks/helpers.py:
```diff
@@ -0,0 +1,235 @@
+    # Let details be None for opening/closing, but preserve old title on edits
+    if action in ['opened', 'closed', 'reopened', 'edited']:
+        changes = payload.get('changes')
+        if changes is not None:
```
You could also just write `if change:`, it will fail on falsey stuff (like `None`).

karlcow replied:
@laghee 

just to be sure. Do you know if there is a "changes" with content when `opened`, `reopened` or `closed` action are happening?

and when `edited` is there any chance that changes is not there?

laghee replied:
@karlcow Nope -- for `opened`, `reopened`, and `closed`, there is no `changes` key. If it's not present on `edited`, however, it would be because of some error (or future configuration change) on GitHub's end (https://developer.github.com/v3/activity/events/types/#issuesevent).

miketaylr commented on ochazuke/webhooks/helpers.py:
```diff
@@ -0,0 +1,235 @@
+
+    When a new issue is opened, we insert it into our issue table, including:
+    - github number (int, 'id')
+    - title (text, 'header')
```
What's the reasoning for using `header` to describe the title? Is title being used somewhere else?

laghee replied:
@miketaylr I had this idea that it would be helpful not to have `label.title`, `milestone.title`, and `issue.title`, because I thought having common property names might get confusing, but that may have just made things... more confusing? 🤷‍♀️ Not at all wedded to it if it's a head-scratcher for everyone else. Github decided that labels have names and milestones and issues have titles (for whatever reason). We could change to follow that if it makes more sense.

miketaylr replied:
Gotcha... I'm not 100% against using `header`, if we add a comment somewhere explaining that a header is an abstraction over a `name` and a `title`, depending on the object. 

Otherwise future me will be confused when I go to the GItHub API docs and find no mentions of headers. :)~

miketaylr commented:
Looks really good! I'll let @karlcow have a peek too.

laghee commented on ochazuke/webhooks/__init__.py:
```diff
@@ -0,0 +1,138 @@
+# License, v. 2.0. If a copy of the MPL was not distributed with this
+# file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+"""Flask Blueprint for our "webhooks" module.webhooks.
```
@miketaylr Rats! Nope, looks like I did something funky with all those git cherry-pick/rebase maneuvers...

laghee replied:
Heh. Nope, looks like that's just fallout from my copied-the-encyclopedia-entry-for-my-report / yes-I'm-old-be-quiet / copy-and-paste approach 🙈  :
<img width="605" alt="screenshot 2018-08-10 02 43 57" src="https://user-images.githubusercontent.com/20972610/43932970-494bc712-9c47-11e8-8432-456d3a153d3d.png">

karlcow commented on ochazuke/webhooks/helpers.py:
```diff
@@ -0,0 +1,235 @@
+    bug = Issue(id=info.get('issue_id'), header=info.get('header'),
+                is_open=True, created_at=info.get('created_at'),
+                milestone_id=milestone.id)
+    # TODO: log failures/errors as well?
```
If there's an issue about that pointing to the code, let's remove the comment.

karlcow commented on ochazuke/webhooks/helpers.py:
```diff
@@ -0,0 +1,235 @@
+from datetime import datetime
+
+from ochazuke import app
+from ochazuke.models import db, Milestone, Label, Issue, Event
```
nit: one line per import :)

karlcow commented on ochazuke/webhooks/helpers.py:
```diff
@@ -0,0 +1,235 @@
+
+def process_label_event_info(payload):
+    """Extract necessary information from webhook for label events."""
+    action = payload.get('action')
```
```python
action = payload['action']
```

payload is a dictionary no?
Basically you can directly extract everything. the get is very practical when you want to default on something else if the deferencing doesn't work.

karlcow commented on ochazuke/webhooks/helpers.py:
```diff
@@ -0,0 +1,235 @@
+def process_label_event_info(payload):
+    """Extract necessary information from webhook for label events."""
+    action = payload.get('action')
+    label_name = payload.get('label')['name']
```
is there any moment where this is not true with the same reasoning than the previous comment.
```python
label_name = payload['label']['name']
```

karlcow commented on ochazuke/webhooks/helpers.py:
```diff
@@ -0,0 +1,235 @@
+    label_name = payload.get('label')['name']
+    prior_name = payload.get('changes').get('name', None)
+    if prior_name is not None:
+        prior_name = payload.get('changes')['name']['from']
```
so if `changes` might be here or not. You could do something like this:

```python
>>> payload = {'label': {'name': 'foobar'}}
>>> if 'changes' in payload:
...     print('do something')
... 
>>> payload = {'label': {'name': 'foobar'}, 'changes': {'name': {'from': 'Old stuff'}}}
>>> if 'changes' in payload:
...     print('do something')
... 
do something
```

which means for the code:

```python
if changes in payload:
    prior_name = payload['changes']['name']['from']
```

karlcow commented on ochazuke/webhooks/helpers.py:
```diff
@@ -0,0 +1,235 @@
+    if action == 'created':
+        label = Label(name=label_name)
+        db.session.add(label)
+    elif (action == 'edited') and name_edited:
```
is there a moment where the action is `edited` and the name has not been edited? :)

karlcow replied:
That's where the tests first would have helped a bit. You know there are a series of possible payloads (your tests), then the function tries to deal with these cases. :)

Let me make a more general comment for this review to untie you of the testing scenario.

laghee commented:
UPDATE: Experimenting with the same webhooks set up on my old local webcompat test-repo, it turns out that GH does some weird stuff with issue events. When milestones/labels are added **while** opening an issue, they still send out separate webhooks for those "additional" events.

Example:
- I opened an issue with a milestone and two labels.
- Not only did the milestone/labels come through in the original `opened` action,
- That **also** triggered one `milestoned` action and **two** `labeled` actions (which... fine, I guess?)
- But **then** the truly bizarre thing is that **each** of the **two separate** `labeled` webhook payloads contains **both** labels. 🤔 😕 

So... odd and dumb that we could potentially perform busywork data updates that aren't actually changing anything. But probably won't break too much. 

This means I need to add some logic for processing multiple labels, though. I had assumed there would only ever be one label per payload. (ahhh, assumptions... `<insert mildly offensive cliche here>`)

laghee commented:
@miketaylr @karlcow I made changes based on discussion and requests:

- changed all non-essential `dict.get` usages to `dict['key']`
- changed all issue `header` usages to `title` to match github (also updated in models.py PR #52)
- refactored the database ops into a single, separate helper (`make_change_to_database`)
- pulled issue-event routing functionality (`update_db`) back into the webhook handler to make testing easier (hopefully)

Going back to hacking on tests with my new resources...
